ITensorGaussianMPS/src/linalg.jl

```diff
@@ -0,0 +1,201 @@
+function gaussian_diag_corr(Λ::Hermitian; pert=nothing)
+  #shift correlation matrix by half so spectrum is symmetric around 0
+  populations, U = gaussian_eigen(Λ - 0.5 * I; pert=pert)
+  n = diag(U' * Λ * U)
+  if !all(abs.(populations - (n - 0.5 * ones(size(n)))) .< sqrt(eps(real(eltype(Λ)))))
+    @show n
+    @show populations .+ 0.5
+    @error(
+      "The natural orbital populations are not consistent, see above. Try adding symmetric noise to the input matrix."
+    )
+  end
+  return populations .+ 0.5, U
+end
+
+"""Takes a single-particle correlation matrix in interlaced Dirac format and finds the fermionic transformation U that diagonalizes it"""
+function gaussian_diag_corr(Γ::AbstractMatrix; pert=nothing)
+  #enforcing hermitianity
+  Γ = (Γ + Γ') / 2.0
+  return gaussian_diag_corr(Hermitian(Γ); pert=pert)
+end
+
+"""Schur decomposition of skew-hermitian matrix"""
+function order_schur(F::LinearAlgebra.Schur)
+  T = F.Schur
+  O = F.vectors #column vectors are Schur vectors
+
+  N = size(T, 1)
+  n = div(N, 2)
+  shuffled_inds = Vector{Int}[]
+  ElT = eltype(T)
+  vals = ElT[]
+  # build a permutation matrix that takes care of the ordering
+  for i in 1:n
+    ind = 2 * i - 1
+    val = T[ind, ind + 1]
+    if real(val) >= 0
+      push!(shuffled_inds, [ind, ind + 1])
+    else
+      push!(shuffled_inds, [ind + 1, ind])
+    end
+    push!(vals, abs(val))
+  end
+  # build block local rotation first
+  perm = sortperm(real.(vals); rev=true)    ##we want the upper left corner to be the largest absolute value eigval pair?
+  vals = vals[perm]
+  shuffled_inds = reduce(vcat, shuffled_inds[perm])
+  # then permute blocks for overall ordering
+  T = T[shuffled_inds, shuffled_inds]
+  O = O[:, shuffled_inds]
+  return T, O, vals #vals are only positive, and of length n and not N
+end
+
+"""Checks if we can make degenerate subspaces of a U0 real by multiplying columns or rows with a phase"""
+function make_real_if_possible(U0::AbstractMatrix, spectrum::Vector; sigdigits=12)
+  # only apply to first half of spectrum due to symmetry around 0
+  # assumes spectrum symmetric around zero and ordered as vcat(-E,E) where E is ordered in descending magnitude
+  U = copy(U0)
+  n = div(length(spectrum), 2)
+  # Round spectrum for comparison within finite floating point precision.
+  # Not the cleanest way to compare floating point numbers for approximate equality but should be sufficient here.
+  rounded_halfspectrum = round.(spectrum[1:n], sigdigits=sigdigits)
+  approx_unique_eigvals = unique(rounded_halfspectrum)
+  # loop over degenerate subspaces
+  for e in approx_unique_eigvals
+    mask = rounded_halfspectrum .== e
+    if abs(e) < eps(real(eltype(U0)))
+      # handle values close to zero separately
+      # rotate subspace for both positive and negative eigenvalue if they are close enough to zero
+      mask = vcat(mask, mask)
+      subspace = copy(U[:, :][:, mask])
```

Comment:
You should be able to just use `subspace = U[:, mask]` here, slicing makes a copy by default (i.e. if you don't use `@view`).